cep-recipe-jinja.md

```diff
@@ -0,0 +1,527 @@
+
+<table>
+<tr><td> Title </td><td> A new recipe format: `jinja` functions in recipes </td>
+<tr><td> Status </td><td> Open </td></tr>
+<tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>
+<tr><td> Created </td><td> Apr 12, 2024</td></tr>
+<tr><td> Updated </td><td> Jun 10, 2024</td></tr>
+<tr><td> Discussion </td><td> https://github.com/conda/ceps/pull/71 </td></tr>
+<tr><td> Implementation </td><td>https://github.com/prefix-dev/rattler-build</td></tr>
+</table>
+
+## Abstract
+
+This CEP is part of the effort to strictly define a new recipe format. The previous CEPs are:
+
+- [CEP 13: A new recipe format](https://github.com/conda/ceps/blob/main/cep-13.md)
+- [CEP 14: A new recipe format: allowed keys and values](https://github.com/conda/ceps/blob/main/cep-14.md)
+
+Historically, `conda-build` recipes have relied on templating with `Jinja` for some
+"dynamic" functionality. For example, many recipes use the `version` of the
+package in multiple places (as package version, in the URL and the tests, for
+example). To make it easy to change recipes, Jinja has been used for some
+light-weight templating.
+
+The "old" recipe format has allowed arbitrary Jinja syntax (including set,
+if/else or for loops). The new recipe format only allows a subset of Jinja with
+the goal of always producing valid YAML files. In this CEP we clarify how Jinja
+is used in the new recipe format, what Jinja functions are available and how
+variables can be set and used.
+
+## Jinja in the new recipe format
+
+The new recipe format uses a subset of Jinja. Specifically, only "variable"
+expressions are allowed (no blocks such as `set`, `for` loops, `if`/`else`
+blocks, ...).
+
+A Jinja expression in the new recipe format looks like the following:
+
+`${{ version }}`
+
+Or if a function is involved:
+
+`${{ compiler('c') }}`
+
+Jinja expressions are also used in `if` statements and in the `skip` field of a
+recipe. In both instances, the `${{ ... }}` syntax is omitted.
+
+E.g.:
+
+```yaml
+build:
+  skip:
+    - osx  # This is a Jinja expression!
+
+requirements:
+  build:
+    - if: win and cuda  # This is a Jinja expression!
+      then:
+        - cudatoolkit
+```
+
+## Variables in the recipe
+
+The variables that are available in the Jinja context in the recipe come from
+two sources: the "variant configuration" file or the "context" section of the
+recipe.
+
+### The context section
+
+The context is a dictionary at the top-level of a recipe that maps keys to
+scalar values. The keys can be used by accessing them as variables in the Jinja
+expressions. For example:
+
+```yaml
+context:
+  version: "1.0.5"
+
+package:
+  version: ${{ version }}
+```
+
+Context evaluation must happen from top-to-bottom. That means a later value can
+reference an earlier one like so:
+
+```yaml
+context:
+  version: "1.0.5"
+  name_and_version: "pkg_${{ version | replace('.', '_') }}"  # evaluates to "pkg_1_0_5"
+```
```

Comment:
```suggestion
```
It would be better to leave this under-specified here instead of specifying it in conflict with the CEP-14.
This exact issue with the specification has been brought up by me in both previous CEPs for the new format and the wording here regresses on the issue instead of finally addressing it.
I do not understand why this simple issue is not being addressed, given that solutions have already been proposed.